### ISSUE TITLE
Training on imdb

### ISSUE THREAD

**Issue by fciannella** [artemisart/bert#1]
What parameters did you use for training on the imdb when running the run_classifier.py? Could you share your configuration? When I run it with your code (imdb branch) with default parameters, I get very low accuracy, like 50%.

I used the default command (omitting the definition of the env variables):

```
python run_classifier.py --task_name=imdb --do_train=true --do_eval=true --data_dir=$BERT_DATA_DIR --vocab_file=$BERT_BASE_DIR/vocab.txt --bert_config_file=$BERT_BASE_DIR/bert_config.json --init_checkpoint=$BERT_BASE_DIR/bert_model.ckpt --max_seq_length=256 --train_batch_size=32 --learning_rate=2e-5 --num_train_epochs=5.0 --output_dir=$TMP_DIR
```

**Comment by artemisart** [comment]
The parameters in train_imdb.sh should work if I remember well (i.e. ~90% acc at least, in 1 epoch), what is your pretrained model ? And try the biggest batch_size you can (and you can reduce seq length to 128 maybe for starting).

**Comment by fciannella** [comment]
Thanks for your reply. I tried with your script:

```
root@fciannel_tf:/tmp/bert# cat train_imdb.sh 
#!/usr/bin/env bash

BERT_BASE_DIR='/home/fciannel/src/bert/bert_base/uncased_L-12_H-768_A-12'
# BERT_BASE_DIR='/data/checkpoints/uncased_L-12_H-768_A-12'
# BERT_BASE_DIR='/data/checkpoints/uncased_L-24_H-1024_A-16'
# DATASETS="$HOME/.data/glue_data"
DATASETS='/tmp'

tensorboard --logdir /data/outputs &

python run_classifier.py \
  --task_name=IMDB \
  --do_train=true \
  --do_eval=true \
  --data_dir=$DATASETS/IMDB \
  --vocab_file=$BERT_BASE_DIR/vocab.txt \
  --bert_config_file=$BERT_BASE_DIR/bert_config.json \
  --init_checkpoint=$BERT_BASE_DIR/bert_model.ckpt \
  --max_seq_length=128 \
  --train_batch_size=32 \
  --learning_rate=2e-5 \
  --num_train_epochs=3 \
  --output_dir=/tmp/imdb_tmp \
  $@
```

But at the end of the training I get the following:

```
INFO:tensorflow:Restoring parameters from /tmp/imdb_tmp/model.ckpt-2343
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2019-01-08-15:50:52
INFO:tensorflow:Saving dict for global step 2343: eval_accuracy = 0.50464, eval_loss = 3.91697, global_step = 2343, loss = 3.9169862
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 2343: /tmp/imdb_tmp/model.ckpt-2343
INFO:tensorflow:***** Eval results *****
INFO:tensorflow:  eval_accuracy = 0.50464
INFO:tensorflow:  eval_loss = 3.91697
INFO:tensorflow:  global_step = 2343
INFO:tensorflow:  loss = 3.9169862
```

So I use the same initial model (uncased_L-12_H-768_A-12) as yours in the script.

Did you change any parameter in bert_config.json, or can you share your log?

I really would like to have this working! Thanks for your help!

**Comment by hsm207** [comment]
I am also getting the same results as @fciannel 

@artemisart in the [bert ](https://github.com/google-research/bert/issues/61)repo you mentioned:

> (with max_seq_length=256 for training -colab limitation- and 512 for evaluation)

What did you mean by this?

**Comment by fciannella** [comment]
I guess he meant that you can do the evaluation as a separate function, like in the colab example. I have also tried that in colab, like @artemisart suggested but still getting the same 50% result :(

Have a look at the google example:

https://colab.research.google.com/github/tensorflow/tpu/blob/master/tools/colab/bert_finetuning_with_cloud_tpus.ipynb

@artemisart  if you can share the way you used to get to 90% it would be very useful!

**Comment by hsm207** [comment]
@fciannel please see my solution [here](https://github.com/google-research/bert/issues/212#issuecomment-453767519)

**Comment by artemisart** [comment]
I'll see if I still have the colab notebooks or logs.

> > (with max_seq_length=256 for training -colab limitation- and 512 for evaluation)
> 
> What did you mean by this?

You can train with max_seq_length=256, save the model, and run the final evaluation with the saved model and with max_seq_length=512 and it will still have a better performance (i.e. significant increase of the metrics), thanks to the pretraining (the last position embeddings were not finetuned because max_seq_length=256 but the model hasn't forgotten them if you will, and you don't ignore words in the 256-512 range).

**Comment by artemisart** [comment]
> @fciannel please see my solution [here](https://github.com/google-research/bert/issues/212#issuecomment-453767519)

I just read it, indeed you must shuffle the data ! You will not be able to learn anything if you first have 12500 neg examples and then 12500 pos, and you can see in the logs if the first 5 examples appear shuffled. The question is then why my code doesn't shuffle the data (anymore?).

**Comment by artemisart** [comment]
@fciannel @hsm207 
Ok so did you use my script create_imdb_dataset.py ? Because I do shuffle the data here, so you shouldn't have had the problem with my script.

As for the logs: some old colab https://colab.research.google.com/drive/1MsCDi6y_4OjJrYbbiBdFC6dKFI2Hv9zt (.9322 eval acc)

**Comment by hsm207** [comment]
@artemisart 
Are you referring to the [create_imdb_dataset.py](https://github.com/artemisart/bert/blob/imdb/create_imdb_dataset.py) in the imdb branch? If yes, then I just ran this script and it did not shuffle the training set. Only the dev and test set got shuffled. I've submitted a pull request to fix it.

**Comment by artemisart** [comment]
Indeed, my bad - I reread it a bit too fast, the train.tsv is not shuffled but you can see in the colab that the examples are shuffled on loading, ~~if I'm not mistaken it's from https://sourcegraph.com/github.com/artemisart/bert@c4c72b5d26f6b3f9680e5e4b12b60d7d5f96f7d8/-/blob/run_classifier.py#L570~~
It shuffles only in a window of 100 examples, so I don't remember why my dataset is shuffled in my colab...
Thanks for the PR, I'll merge it !
